### DISCUSSION TITLE
Running phpunit with WordPress (and testing a plugin)

### DISCUSSION THREAD

**Discussion by Ebeldev** [ddev/ddev#4459] [General]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Run a Diagnostic and Paste Link Here

I wanted to know if there is a way to setup or guide on how to setup a WordPress plugin to use the phpunit with DDEV ?

So far my reasearch and tests have not been successfull. 

Thank you for your help, 

### Expected Behavior

I want to use the wp cli scaffold command with the use of their script to install and use phpunit. 

But it does not work. 

### Actual Behavior

I want to setup a WordPress plugin to include phpunit tests. 

### Anything else?

I wanted to know if there is a way to setup or guide on how to setup a WordPress plugin to use the phpunit with DDEV ?

So far my reasearch and tests have not been successfull. 

Thank you for your help,

**Comment by rfay** [comment]
Do you already have the phpunit tests in there? Have you composer-installed phpunit?

**Comment by Ebeldev** [comment]
Hi @rfay , 

Yes, I have installed phpunit on my machine with homebrew. And I also installed it, locally in the plugin.

**Comment by Ebeldev** [comment]
![Screen Shot 2022-11-15 at 1 18 28 PM](https://user-images.githubusercontent.com/8010184/201995993-cb217daf-4f9a-47ee-b03c-54cc3b232510.png)

**Comment by rfay** [comment]
You need phpunit in your code's vendor directory. Install it with composer.  Having it on your computer is not relevant. The code in the container cannot access anything on your computer. 

Please use `ddev ssh` to start learning about this. When you can use `phpunit` in there, then you are ready to start.

**Comment by rfay** [comment]
phpunit when installed is usually in vendor/bin/phpunit, for example. see https://github.com/weitzman/ddev-selenium-standalone-chrome and its readme.

**Comment by rfay** [comment]
@mandrasch can you offer any help here? (@mandrasch is an expert with WP and DDEV)

**Comment by Ebeldev** [comment]
@rfay I forgot to mentionned it, I also did try it in ssh in the container. 

Here is the error when I then start my unit tests. 

mac-dw-eti@rabaischocs-web:/var/www/html/wp-content/plugins/rc-reports$ ./vendor/bin/phpunit 
Could not find /tmp/wordpress-tests-lib/includes/functions.php, have you run bin/install-wp-tests.sh ?

I did follow the setup instruction : https://pippinsplugins.com/unit-tests-wordpress-plugins-setting-up-testing-suite/.  you set up wordpress with phpunit, with the wp cli it uses a bash script. After the script has been run. And then try to do any unit test I get the error mentionned. 

From our understanding the error would be in the bash script. The script seems to not know it's in a container.

**Comment by Ebeldev** [comment]
@rfay By the way, I really like working with ddev.

**Comment by rfay** [comment]
Have you run bin/install-wp-tests.sh? 

It seems to install something in /tmp/wordpress-tests-lib, and /tmp is a volatile directory in ddev (it gets recreated on every start)

**Comment by Ebeldev** [comment]
Yes I did run bin/install-wp-tests.sh command.

**Comment by rfay** [comment]
You'll need to run it each time you restart, since /tmp/ goes away on restart.

You need to look in /tmp/wordpress-tests-lib and see what's there or not there. 

Make sure everything you're doing right now is inside the container (`ddev ssh`)

Lots of other people have had this problem before you, https://www.google.com/search?q=Could+not+find+%2Ftmp%2Fwordpress-tests-lib%2Fincludes%2Ffunctions.php%2C+have+you+run+bin%2Finstall-wp-tests.sh&oq=Could+not+find+%2Ftmp%2Fwordpress-tests-lib%2Fincludes%2Ffunctions.php%2C+have+you+run+bin%2Finstall-wp-tests.sh&aqs=chrome..69i57j0i512.239j0j9&sourceid=chrome&ie=UTF-8

**Comment by rfay** [comment]
No response after quite some time, I'll convert this to a discussion.

**Comment by Ebeldev** [comment]
@rfay 
I know it's been a while. 

I was able to run the ddev ssh, and run the WordPress scaffold command with wp-cli. 

But I still have issues. 

I wish there were a guide on how to setup PhpUnit with WordPress on DDEV. 

Unfortunatly, so far all my tests have gone cold. 

@mandrasch Have you ever setup a PhpUnit in a WordPress plugin ?

**Comment by Ebeldev** [comment]
I will look into using wp-browser to run all my tests. I'll let you know if I was able to connect this.

**Comment by Ebeldev** [comment]
@rfay I was able to install wp-browser. 

I'm following the VVV setup, because VVV works with Containers, from my understanding.

I'm trying to create database in mysql. With those 2 commands : 
mysql -u root -p -e "CREATE DATABASE if not exists tests"
mysql -u root -p -e "GRANT ALL PRIVILEGES ON tests.* TO 'wp'@'localhost';"

And it's asking me my password. Which password should it be ? I tried, my computer one, the db password and no password. 
I always get a access denied. 
![Screenshot 2023-06-15 at 9 05 57 PM](https://github.com/ddev/ddev/assets/8010184/2982efa8-58a1-4433-8307-65a110ce093e)

Any idea for the password ?

**Reply by rfay** [reply]
`ddev describe` tells you this always. But the database is named `db` and the user is `db` (not wp) and the host is `db` not localhost, and the password is `db`.

**Reply by Ebeldev** [reply]
Yes I did find the ddev describe. Thanks.

I did change the database name to 'db'. 
Still the db password do not work. And I even tried the root password root. Even there no luck.

**Reply by Ebeldev** [reply]
ok using root worked !

**Reply by rfay** [reply]
You'll want to see the FAQ, https://ddev.readthedocs.io/en/latest/users/usage/faq/#how-can-i-connect-to-my-database

And https://ddev.readthedocs.io/en/latest/users/usage/database-management/

**Reply by Ebeldev** [reply]
I was able to install wp-browser and setup correct all. 

But when I try to do the basic 4 tests.Only unit tests works. 

Acceptance, functionnal and wpunit do have errors. 
![Screenshot 2023-06-15 at 10 48 32 PM](https://github.com/ddev/ddev/assets/8010184/075ef6c2-bea1-4793-902e-6fe30be866b6)

**Reply by rfay** [reply]
Your tests still aren't using the right db credentials it looks like. "No such file or directory creating PDO"...

**Reply by Ebeldev** [reply]
I was able to create the tests database using user root and password root. 

When I try to debug :
vendor/bin/codecept run acceptance --debug

I get : 
[Connecting To Db] {"config":{"tablePrefix":"wp_","populate":true,"cleanup":true,"reconnect":false,"dump":"tests/_data/dump.sql","populator":null,"urlReplacement":true,"originalUrl":null,"waitlock":10,"dsn":"mysql:host=localhost;dbname=tests","user":"root","password":"root","url":"https://wpcleantest.ddev.site"},"options":[]}

even if I change the user and credentials of the db i get a error for the connection : 
[Connecting To Db] {"config":{"tablePrefix":"wp_","populate":true,"cleanup":true,"reconnect":false,"dump":"tests/_data/dump.sql","populator":null,"urlReplacement":true,"originalUrl":null,"waitlock":10,"dsn":"mysql:host=localhost;dbname=tests","user":"db","password":"db","url":"https://wpcleantest.ddev.site"},"options":[]}

I also tried : 
mysql -u db -p db -h localhost -P 8036 -D tests -e "show tables";
mysql -u root -p root -h localhost -P 8036 -D tests -e "show tables";

Both of those gives me a error. 

When I open the phpmyadmin: 
I see the table
![Screenshot 2023-06-16 at 7 12 12 AM](https://github.com/ddev/ddev/assets/8010184/d9cb031f-654f-4869-b7c4-13c39e073926)

I used thoses two lines to create the db:
mysql -u root -p -e "CREATE DATABASE if not exists tests"
mysql -u root -p -e "GRANT ALL PRIVILEGES ON tests.* TO 'db'@'localhost';"
![Screenshot 2023-06-16 at 7 28 04 AM](https://github.com/ddev/ddev/assets/8010184/def54e10-2f3e-4b84-b013-c1e353c131bf)

... 

I did more tests. 

And I think my problem is that I do not connect the good way when i'm using the command line once in the docker ssh. 
I tried again the 2 lines above, and they do not work this morning. 
But when I using ddev mysql -uroot -proot, i'm able to connect to the db and add database.

**Reply by rfay** [reply]
Port 8036 is used for PhpMyAdmin. You don't have any need to be using it. The database port is 3306. 

To access the database inside the web container or DB container, `mysql` will do the job. Or `mysql -uroot -proot`. Or `mysql -uroot -proot db`  Again, `ddev describe` shows you these things.

To access the database from the host (workstation) side, `ddev mysql` works nicely. Same args.

**Reply by Ebeldev** [reply]
@rfay 
Thanks. 

I'm able to see my databases and connect in the mysql with the command line. 
mysql -uroot -proot -e "show databases;"
But still the acceptance tests with wpbrowser cant connect to the db. 

I'm in contact as well with the wp-browser to see why it's not connecting. 

I followed this guide. Everything is pretty straight forward. 
https://wpbrowser.wptestkit.dev/tutorials/vvv-setup

I'm in the last step: sanity check
vendor/bin/codecept run acceptance
vendor/bin/codecept run functional
vendor/bin/codecept run wpunit
vendor/bin/codecept run unit

Only the unit test connects correctly. 

Here is my .env.testing constants file. 
TEST_SITE_DB_DSN=mysql:host=localhost;dbname=tests
TEST_SITE_DB_HOST=localhost
TEST_SITE_DB_NAME=tests
TEST_SITE_DB_USER=root
TEST_SITE_DB_PASSWORD=root
TEST_SITE_TABLE_PREFIX=wp_
TEST_SITE_ADMIN_USERNAME=admin
TEST_SITE_ADMIN_PASSWORD=admin
TEST_SITE_WP_ADMIN_PATH=/wp-admin
WP_ROOT_FOLDER=/var/www/html
TEST_DB_NAME=tests
TEST_DB_HOST=localhost
TEST_DB_USER=root
TEST_DB_PASSWORD=root
TEST_TABLE_PREFIX=wp_
TEST_SITE_WP_URL=https://wpcleantest.ddev.site
TEST_SITE_WP_DOMAIN=wpcleantest.ddev.site
TEST_SITE_ADMIN_EMAIL=dev@groupejcl.ca


![Screenshot 2023-06-17 at 8 50 08 AM](https://github.com/ddev/ddev/assets/8010184/a6d7f293-014c-4399-8c12-b785b0ef58d8)

I'll keep searching. 

I know the scope of this is out of ddev.

**Reply by rfay** [reply]
You're accidentally trying to use a database called "root". You have your stuff in the wrong place. "Unknown database 'root'"

**Reply by Ebeldev** [reply]
Yes in the wrong call with mysql. 

But with the good call with mysql, i do see the database. But the codecept doesn't seems to be able to connect. 

Yes! 
I finally made Acceptance and Functionnal tests works. 
I replace the host from localhost to db. And it worked. 

Sorry for this all. 

Thanks for your replies.

**Reply by rfay** [reply]
I'm glad you made it work. Could you please summarize here how to make tests work with WordPress? Others will be interested.

**Comment by Ebeldev** [comment]
Thanks

**Comment by Ebeldev** [comment]
@rfay 

Hi Randy, 

Here is my summary on the steps to setup DDEV with WP-Browser. 

I did follow this guide https://wpbrowser.wptestkit.dev/tutorials/vvv-setup. 

I just made a few changes to be able to to build on DDEV.

1. Make sure your ddev project runs. At the root of your project, do this commands `ddev describe`
2. Made a new plugin or a theme with the bare minimum of WordPress to work. 
3. Ssh into Ddev container : `ddev ssh`
4. Once in the container, `cd wp-content/plugins` in the plugin or theme directory
5. `composer init`
6. Download WP-Browser pakage : `composer require --dev lucatume/wp-browser`
7. Make sure Wp-Browser and Codeception is correctly install : `vendor/bin/codecept --version`
8. Create a dummy tests database in mysql with the next 2 commands
9. `mysql -u root -p -e "CREATE DATABASE if not exists tests"` password = root
10. `mysql -u root -p -e "GRANT ALL PRIVILEGES ON tests.* TO 'db'@'localhost';"` passowrd = root
11. Check the database was correctly created running this command: `mysql -u root -p -e "SHOW DATABASES"`
12. If your already in your plugin or theme directory, configure WP-Browser with this command : `vendor/bin/codecept init wpbrowser`
13. Then for all questions, I followed the questions of the wizard. Except for the localhost. For DDEV projects : ddev-nameofyourproject-db
14. Not sure of your localhost. Look in your wp-config.php file. Not sure of what to put. No worries you can always go back after in your plugin or theme project folder and look for the file .env.testing. I'll put a copy of mine below. 
15. Setup a starting point database for tests. 
16. Again I did this in a clean new WordPress site. 
17. I cd back to the root of WordPress 
`cd ../../
wp plugin deactivate --all
wp plugin activate my-plugin
wp db export wp-content/plugins/my-plugin/tests/_data/dump.sql
`
18. I edited the wp-config-ddev.php file 
change thoses lines : 
define( 'DB_NAME', 'db' );
/** MySQL database username */
defined( 'DB_USER' ) || define( 'DB_USER', 'db' );

/** MySQL database password */
defined( 'DB_PASSWORD' ) || define( 'DB_PASSWORD', 'db' );

for thoses lines
`
if( isset( $_SERVER['HTTP_X_TEST_REQUEST'] ) && $_SERVER['HTTP_X_TEST_REQUEST'] ) {
    define( 'DB_NAME', 'tests' );
		/** MySQL database username */
		defined( 'DB_USER' ) || define( 'DB_USER', 'root' );

		/** MySQL database password */
		defined( 'DB_PASSWORD' ) || define( 'DB_PASSWORD', 'root' );
	} else {
		define( 'DB_NAME', 'db' );
		/** MySQL database username */
		defined( 'DB_USER' ) || define( 'DB_USER', 'db' );

		/** MySQL database password */
		defined( 'DB_PASSWORD' ) || define( 'DB_PASSWORD', 'db' );
	}
`
19. Sanity check 
`
vendor/bin/codecept run acceptance
vendor/bin/codecept run functional
vendor/bin/codecept run wpunit
vendor/bin/codecept run unit
`
All tests should be good. 
If you have a problem check your variables in the .env.testing file. 

You can use any of the vendor/bin/codecept run ... commands with --debug to have more details if you have a problem. 
example : `vendor/bin/codecept run acceptance --debug`


// .env.testing example
TEST_SITE_DB_DSN=mysql:host=ddev-wpcleantest-db;dbname=tests
TEST_SITE_DB_HOST=ddev-wpcleantest-db
TEST_SITE_DB_NAME=tests
TEST_SITE_DB_USER=root
TEST_SITE_DB_PASSWORD=root
TEST_SITE_TABLE_PREFIX=wp_
TEST_SITE_ADMIN_USERNAME=admin
TEST_SITE_ADMIN_PASSWORD=admin
TEST_SITE_WP_ADMIN_PATH=/wp-admin
WP_ROOT_FOLDER=/var/www/html
TEST_DB_NAME=tests
TEST_DB_HOST=ddev-wpcleantest-db
TEST_DB_USER=root
TEST_DB_PASSWORD=root
TEST_TABLE_PREFIX=wp_
TEST_SITE_WP_URL=https://wpcleantest.ddev.site
TEST_SITE_WP_DOMAIN=wpcleantest.ddev.site
TEST_SITE_ADMIN_EMAIL=admin@example.com

**Reply by Ebeldev** [reply]
Before step 19. 

Cd back into your plugin or theme you installed composer earlier. Before doing the sanity check

**Reply by rfay** [reply]
Awesome work. (You can edit your previous post if you need to add more.)

**Comment by Ebeldev** [comment]
All tests works. 

So I'm ready to write my first test. 

![Screenshot 2023-06-18 at 9 38 34 PM](https://github.com/ddev/ddev/assets/8010184/6e72ae40-5128-4ca8-8a46-aac72e806060)

**Reply by Ebeldev** [reply]
Sanity check his all good

**Comment by Ebeldev** [comment]
Need help on writing your first Acceptance test : https://wp-punk.com/simple-start-with-acceptance-testing-for-wordpress/

**Comment by Ebeldev** [comment]
@rfay 
If we want to install a specific chrome driver, normally I would do someting like `brew cask install chromedriver`.

How can we do that in a container based system like DDEV ?

**Comment by tyler36** [comment]
[ddev-selenium-standalone-chrome](https://github.com/ddev/ddev-selenium-standalone-chrome) can be used for running Chrome with PHPUnit test inside DDEV on Drupal. It might give you hints.

There are also a couple of other addons you can find by running: `ddev get --list --all`

- [thunder/ddev-selenium-chrome](https://github.com/thunder/ddev-selenium-chrome)

Or try DDEV-contrib

- https://github.com/ddev/ddev-contrib/blob/master/docker-compose-services/headless-chrome/README.md
